internal/controllers/watcher/controller.go

```diff
@@ -0,0 +1,173 @@
+	wm.logger.Info("Starting watcher manager")
+	go func() {
+		err := (*wm.mgr).Start(ctx)
+		if err != nil {
+			wm.logger.Error(err, "there was an error while running the reconciliation watcher")
+		}
+	}()
+}
+
+// WatcherReconciler triggers reconciliation in the AddonManager when something changes in an upstream resource
+type WatcherReconciler struct {
+	client.Client
+	Log           logr.Logger
+	Scheme        *runtime.Scheme
+	addonnManager *addonmanager.AddonManager
+}
+
+var (
+	noNameErr      = errors.New("no name for reconciliation request")
+	noNamespaceErr = errors.New("no namespace for reconciliation request")
+)
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if req.Name == "" {
+		return ctrl.Result{}, noNameErr
+	}
+	if req.Namespace == "" {
+		return ctrl.Result{}, noNamespaceErr
+	}
+	(*r.addonnManager).Trigger(req.Namespace, req.Name)
+
+	r.Log.V(2).Info("reconciliation triggered", "cluster", req.Namespace)
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *WatcherReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&addonapiv1alpha1.ManagedClusterAddOn{}, noReconcilePred).
+		Watches(&corev1.Secret{}, r.enqueueForClusterSpecificResource(), builder.OnlyMetadata).
+		Watches(&loggingv1.ClusterLogForwarder{}, r.enqueueForClusterSpecificResource(), builder.OnlyMetadata).
+		Watches(&otelv1alpha1.OpenTelemetryCollector{}, r.enqueueForClusterSpecificResource(), builder.OnlyMetadata).
+		Complete(r)
+}
+
+func (r *WatcherReconciler) enqueueForClusterSpecificResource() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []reconcile.Request {
+		key := client.ObjectKey{Name: "multicluster-observability-addon", Namespace: obj.GetNamespace()}
+		addon := &addonapiv1alpha1.ManagedClusterAddOn{}
+		if err := r.Client.Get(ctx, key, addon); err != nil {
+			if apierrors.IsNotFound(err) {
+				switch obj.(type) {
+				case *corev1.Secret:
+					return r.getSecretReconcileRequests(ctx, obj, addon)
+				}
+				return nil
+			}
+			r.Log.Error(err, "Error getting managedclusteraddon resources in event handler")
+			return nil
+		}
+
+		return []reconcile.Request{
+			{
+				NamespacedName: types.NamespacedName{
+					Name:      addon.Name,
+					Namespace: addon.Namespace,
+				},
+			},
+		}
+	})
+}
+
+// getSecretReconcileRequests gets reconcile.Request for secrets referenced in ManifestWorks.
+func (r *WatcherReconciler) getSecretReconcileRequests(ctx context.Context, obj client.Object, addon *addonapiv1alpha1.ManagedClusterAddOn) []reconcile.Request {
+	rqs := []reconcile.Request{}
+	mws := &workapiv1.ManifestWorkList{}
+	if err := r.Client.List(ctx, mws); err != nil {
```

Comment:
Here we should list only the ManifestWork that belong to MCOA we can do so in the following way:
```suggestion
	if err := r.Client.List(ctx, mws, &client.ListOptions{
		LabelSelector: labels.SelectorFromSet(labels.Set{
			"open-cluster-management.io/addon-name": "multicluster-observability-addon",
		}),
	}); err != nil {
```